src/common/crypto/sha512.cpp

```diff
@@ -0,0 +1,35 @@
+#include "duckdb/common/crypto/sha512.hpp"
+
+namespace duckdb {
+
+SHA512Context::SHA512Context() {
+	mbedtls_sha512_init(&sha_context);
+	mbedtls_sha512_starts(&sha_context, false);
+}
+
+SHA512Context::~SHA512Context() {
+	mbedtls_sha512_free(&sha_context);
+}
+
+void SHA512Context::Add(string_t str) {
+	mbedtls_sha512_update(&sha_context, (const unsigned char *)str.GetData(), str.GetSize());
+}
+
+void SHA512Context::Finish(char *out) {
+	unsigned char digest[SHA512_HASH_LENGTH_BINARY];
+	mbedtls_sha512_finish(&sha_context, digest);
+	DigestToBase16(digest, out);
+}
+
+void SHA512Context::DigestToBase16(const_data_ptr_t digest, char *zbuf) {
+	static char const HEX_CODES[] = "0123456789abcdef";
+	int i, j;
+
+	for (j = i = 0; i < 64; i++) {
```

Comment:
Yeah. That's a nice idea. Let's talk about how to unify them.
 How about making `CryptoContext` a virtual base class of the `MD5Context`/`SHA512Context`? The `CryptoContext` defines common interfaces.

Reply:
Static methods cannot be virtual - so I would just say have the `CryptoContext` contain both MD5 and SHA methods

Reply:
I have no idea about how to implement the method `CryptoContext::Add`. Since we know the final hash function until all inputs are added to the context, should the `CryptoContext` keep a buffer to cache all inputs in the process?

Reply:
Hm - yes you are correct. Perhaps it would be cleaner to have inheritance here, just without virtual methods (i.e. `DigestToBase16` could be a static method in `CryptoContext` that is then used by both `MD5Context` and `Sha512Context`).

Reply:
If we add an abstract class `CryptoContext`, it seems that we would never refer to it. It will just be an `interface`.   
Since the `DigestToBase16` is not only used in `MD5Context`/`SHA512Context`, how about adding a `HexUtil` like `StringUtil` and moving the `DigestToBase16` in?